File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0437_Path_Sum_III.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.HashMap;
import java.util.Map;

public class Question_0437_Path_Sum_III {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private int res = 0;

    public int pathSum(TreeNode root, int targetSum) {
        res = 0;
        if (root == null) {
            return res;
        }
        Map<Long, Integer> map = new HashMap<>();
        map.put(0L, 1);
        help(root, targetSum, 0, map);
        return res;
    }

    private void help(TreeNode root, int targetSum, long cur, Map<Long, Integer> map) {
        if (root == null) {
            return;
        }
        cur += root.val;
        if (map.containsKey(cur - targetSum)) {
            res += map.get(cur - targetSum);
        }
        if (map.containsKey(cur)) {
            map.put(cur, map.get(cur) + 1);
        } else {
            map.put(cur, 1);
        }
        help(root.left, targetSum, cur, map);
        help(root.right, targetSum, cur, map);
        map.put(cur, map.get(cur) - 1);
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_0437_Path_Sum_III.cc
#include <unordered_map>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  int pathSum(TreeNode* root, int targetSum) {
    if (!root) { 
      return 0;
    }
    int num = 0;
    unordered_map<long long, long long> num_paths;
    num_paths[0] = 1;
    help(root, targetSum, 0, num, num_paths);
    return num;
  }

  void help(TreeNode* root, int sum, long long cursum, int& num, unordered_map<long long, long long>& num_paths) {
    if (!root) {
      return;
    }
    cursum += root->val;
    if(num_paths.count(cursum - sum)){
      num += num_paths[cursum - sum];
    }
    if(num_paths.count(cursum)){
      num_paths[cursum] += 1;
    } else {
      num_paths[cursum] = 1;
    }
    help(root->left, sum, cursum, num, num_paths);
    help(root->right, sum, cursum, num, num_paths);
    num_paths[cursum] -= 1;
    return;
  }
};
